### PR TITLE
Add @userschema endpoint for getting the user schema.

### PR DESCRIPTION
**Pull request by sneridagh**
Related to #706

### CODE REVIEW

**Comment by sneridagh** [comment]
@jenkins-plone-org please run jobs

**Comment by sneridagh** [comment]
This is the initial work for adding support to get the schema of zc3.forms (#480), that's why the directory is just called schema, for later add the forms support. I will draft the documentation first for such endpoint.

**Comment by sneridagh** [comment]
@jenkins-plone-org please run jobs

**Comment by sneridagh** [comment]
Talked about the @schema endpoint with @tisto, agreed on leave it in the fridge for now, since if we have to define the form in the backend is the same effort than we directly customize it in the frontend. Then still, we will have to define the widget mapping (and overrides for specific forms, if any) somehow. So the pros of having it might be very little.

**Comment by sneridagh** [comment]
@jenkins-plone-org please run jobs

**Comment by erral** [comment]
We need this `@userschema` endpoint in a project, we can work on updating this PR, at least to have the endpoint working.

**Comment by tisto** [comment]
@erral awesome!

**Comment by erral** [comment]
@jenkins-plone-org please run jobs

**Comment by erral** [comment]
We have checked the endpoint and it works as expected, it returns the correct jsonschema for the user profile schema.

I have rebased the branch and run black.

**Comment by erral** [comment]
@jenkins-plone-org please run jobs

**Comment by erral** [comment]
@jenkins-plone-org please run jobs

File: src/plone/restapi/services/userschema/configure.zcml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    xmlns="http://namespaces.zope.org/zope"
+    xmlns:plone="http://namespaces.plone.org/plone"
+    xmlns:zcml="http://namespaces.zope.org/zcml">
+
+    <plone:service
+      method="GET"
+      name="@userschema"
+      for="Products.CMFCore.interfaces.ISiteRoot"
+      factory=".user.UserSchemaGet"
+      permission="zope2.View"
```
Comment (avoinea) [inline src/plone/restapi/services/userschema/configure.zcml:11]: Is this supposed to be public?

**Comment by erral** [comment]
At least we need it when registration for the site is opened (https://demo.plone.org/register), in other cases we could check for the user to be logged-in.

File: src/plone/restapi/services/userschema/configure.zcml (hunk 1 above)
Comment (erral) [reply src/plone/restapi/services/userschema/configure.zcml:11]: At least we need it when registration for the site is opened (https://demo.plone.org/register), in other cases we could check for the user to be logged-in.

**Comment by erral** [comment]
@jenkins-plone-org please run jobs

**Comment by erral** [comment]
@jenkins-plone-org please run jobs

**Review by nileshgulia1 — APPROVED** [review]

**Comment by erral** [comment]
@jenkins-plone-org please run jobs

**Review by tiberiuichim — APPROVED** [review]

**Review by nileshgulia1 — CHANGES_REQUESTED** [review]
I found an issue with date-time serialization when adding a field "date_time" into schema.
```
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 162, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 371, in publish_module
  Module ZPublisher.WSGIPublisher, line 266, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module plone.rest.service, line 22, in __call__
  Module plone.restapi.services, line 22, in render
  Module json, line 234, in dumps
  Module json.encoder, line 201, in encode
  Module json.encoder, line 431, in _iterencode
  Module json.encoder, line 405, in _iterencode_dict
  Module json.encoder, line 438, in _iterencode
  Module json.encoder, line 179, in default
TypeError: Object of type DateTime is not JSON serializable
2022-01-28 22:59:10,975 ERROR   [Zope.SiteErrorLog:35][waitress-3] TypeError: http://localhost:3000/@users/admin
```

**Comment by erral** [comment]
@nileshgulia1 but this error is related to the `@users` endpoint and not to `@userschema` right? I am adding a test with different field types and the user schema information is rendered correctly. Perhaps what we need is a [more complete `@users` endpoint test](https://github.com/plone/plone.restapi/blob/master/src/plone/restapi/tests/test_services_users.py#L37) that checks all possible field types. Right now it only tests basic stuff.

**Comment by erral** [comment]
I have checked it and indeed the user serializer does not take into account datetime objects: https://github.com/plone/plone.restapi/blob/master/src/plone/restapi/serializer/user.py#L50

**Comment by nileshgulia1** [comment]
@erral Indeed, its with @users endpoint. Apart from date_time, some of the existing member fields like "portrait"  currently does not work well with volto.

**Comment by nileshgulia1** [comment]
The other fields work well, I think we can merge it then and take care of other tasks/bugs in subsequent PRs.

**Comment by erral** [comment]
And there is no deserializer of users, the property set is done manually without deserializing the user: https://github.com/plone/plone.restapi/blob/master/src/plone/restapi/services/users/update.py#L104

**Comment by erral** [comment]
> The other fields work well, I think we can merge it then and take care of other tasks/bugs in subsequent PRs.

Yes I do agree. I have some changes and tests for the `@users` endpoint, and we can discuss it further there.

**Review by nileshgulia1 — APPROVED** [review]

**Comment by erral** [comment]
@jenkins-plone-org please run jobs

**Review by avoinea — APPROVED** [review]

**Comment by sneridagh** [comment]
LGTM!

@tisto could you please take a look? Since I opened it I cannot approve it :) 

Then we could merge https://github.com/plone/volto/pull/2917 and release it on 15

**Comment by sneridagh** [comment]
@jenkins-plone-org please run jobs

**Comment by sneridagh** [comment]
Ok, another attempt to merge this one. @tisto I think all is good enough to have this one in. Could you please take a look? Since I opened it I cannot approve it :)

Then we could merge https://github.com/plone/volto/pull/2917

@stevepiercy the Netlify build is red? Could you please take a look?

**Comment by sneridagh** [comment]
@stevepiercy uh I got what happened. I didn't know that we move to md in here! Solved!

File: docs/source/endpoints/userschema.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    "description": "Given a user in the site one can get its properties and available fields."
+    "property=og:description": "Given a user in the site one can get its properties and available fields."
+    "property=og:title": "User schema"
+    "keywords": "Plone, plone.restapi, REST, API, Users, profile"
+---
+
+# User schema
+
+.. note::
+    This is only available on Plone 5.
```
Comment (stevepiercy) [inline docs/source/endpoints/userschema.md:13]:
````suggestion
```{note}
    This is only available on Plone 5.
```
````

File: docs/source/endpoints/userschema.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  html_meta:
+    "description": "Given a user in the site one can get its properties and available fields."
+    "property=og:description": "Given a user in the site one can get its properties and available fields."
+    "property=og:title": "User schema"
+    "keywords": "Plone, plone.restapi, REST, API, Users, profile"
+---
+
+# User schema
+
+.. note::
```
Comment (stevepiercy) [inline docs/source/endpoints/userschema.md:12]:
```suggestion
```

File: docs/source/endpoints/userschema.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    This is only available on Plone 5.
+
+Users in Plone have a set of properties defined by a default set of fields like `fullname`, `email`, `portrait`, etc.
```
Comment (stevepiercy) [inline docs/source/endpoints/userschema.md:15]:
```suggestion
Users in Plone have a set of properties defined by a default set of fields such as `fullname`, `email`, `portrait`, and so on.
```

File: docs/source/endpoints/userschema.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+These properties define the site user's profile and the user itself via the Plone UI or the site managers can add them in a variety of ways (eg. using PAS plugins.)
```
Comment (stevepiercy) [inline docs/source/endpoints/userschema.md:16]:
```suggestion
These properties define the site user's profile and the user itself via the Plone UI, or the site managers can add them in a variety of ways including PAS plugins.
```

File: docs/source/endpoints/userschema.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+These fields are dynamic and customizable by integrators so they do not adhere to a fixed schema interface.
+This dynamic schema is exposed by this endpoint in order to build the user's profile form.
+
+## Getting the user schema
+
+To get the current user schema make a request to the `/@userschema` endpoint::
```
Comment (stevepiercy) [inline docs/source/endpoints/userschema.md:23]:
```suggestion
To get the current user schema, make a request to the `/@userschema` endpoint.
```

File: docs/source/endpoints/userschema.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
+---
+myst:
+  html_meta:
+    "description": "Given a user in the site one can get its properties and available fields."
```
Comment (stevepiercy) [inline docs/source/endpoints/userschema.md:4]:
```suggestion
    "description": "Given a user in the site, one can get its properties and available fields."
```

File: docs/source/endpoints/userschema.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    "property=og:description": "Given a user in the site one can get its properties and available fields."
```
Comment (stevepiercy) [inline docs/source/endpoints/userschema.md:5]:
```suggestion
    "property=og:description": "Given a user in the site, one can get its properties and available fields."
```

File: docs/source/endpoints/userschema.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
+---
```
Comment (stevepiercy) [inline docs/source/endpoints/userschema.md:1]: This file needs to be included in `endpoints/index.md` as a `toctree` entry after `users`.

Here is its current preview:
https://deploy-preview-707--plone-restapi.netlify.app/endpoints/userschema.html

File: docs/source/endpoints/userschema.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+## Getting the user schema
+
+To get the current user schema make a request to the `/@userschema` endpoint::
+
+```{eval-rst}
+..  http:example:: curl httpie python-requests
+    :request: ../../src/plone/restapi/tests/http-examples/userschema.req
+```
+
+The server will respond with the user schema::
```
Comment (stevepiercy) [inline docs/source/endpoints/userschema.md:30]:
```suggestion
The server will respond with the user schema.
```

File: docs/source/endpoints/userschema.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+```{literalinclude} ../../src/plone/restapi/tests/http-examples/userschema.resp
+   :language: http
+```
+
+The user schema uses the same serialization as the types JSON schema one.
```
Comment (stevepiercy) [inline docs/source/endpoints/userschema.md:36]:
```suggestion
The user schema uses the same serialization as the type's JSON schema.
```

File: docs/source/endpoints/userschema.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+See {ref}`types-schema` for a detailed documentation about the available field types.
```
Comment (stevepiercy) [inline docs/source/endpoints/userschema.md:38]:
```suggestion
See {ref}`types-schema` for detailed documentation about the available field types.
```

File: docs/source/endpoints/userschema.md [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+To get the current user schema make a request to the `/@userschema` endpoint::
+
+```{eval-rst}
+..  http:example:: curl httpie python-requests
+    :request: ../../src/plone/restapi/tests/http-examples/userschema.req
+```
+
+The server will respond with the user schema::
+
+```{literalinclude} ../../src/plone/restapi/tests/http-examples/userschema.resp
```
Comment (stevepiercy) [inline docs/source/endpoints/userschema.md:32]: This does not appear in the preview.

https://deploy-preview-707--plone-restapi.netlify.app/endpoints/userschema.html

Try this:

```suggestion
```{literalinclude} ../../../src/plone/restapi/tests/http-examples/userschema.resp
```

File: docs/source/endpoints/userschema.md [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+These fields are dynamic and customizable by integrators so they do not adhere to a fixed schema interface.
+This dynamic schema is exposed by this endpoint in order to build the user's profile form.
+
+## Getting the user schema
+
+To get the current user schema make a request to the `/@userschema` endpoint::
+
+```{eval-rst}
+..  http:example:: curl httpie python-requests
+    :request: ../../src/plone/restapi/tests/http-examples/userschema.req
```
Comment (stevepiercy) [inline docs/source/endpoints/userschema.md:27]: I am surprised that this actually shows up in the preview. Double-check the path.

```suggestion
    :request: ../../../src/plone/restapi/tests/http-examples/userschema.req
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
A few more English and MyST syntax fixes.

File: docs/source/endpoints/userschema.md (hunk 9 above)
Comment (stevepiercy) [reply docs/source/endpoints/userschema.md:1]: This one was missed. Please do this, and then I approve. Thank you!

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Please add the toctree entry to `endpoints/index.md`.

**Review by stevepiercy — APPROVED** [review]
🎉

**Comment by nileshgulia1** [comment]
@jenkins-plone-org please run jobs

**Pull request merged**